### DISCUSSION TITLE
Deprecate environment variable `$PREFIX` and start looking at `$NODE_PREFIX` or `$NPM_PREFIX` to reduce collision potential

### DISCUSSION THREAD

**Discussion by martin-braun** [community/community#128434] [npm]
`nvm` makes use of `$PREFIX` to maintain multiple `node`/`npm` systems, however Termux (a terminal emulator for Android) also uses `$PREFIX` in their environment, so Termux and `nvm` collide. Deprecating the variable on Termux shall be possible, but a lot of third party scripts are out there that use `$PREFIX` to work within Termux.

The idea is to patch `npm` (and `node`) to look for `$NODE_PREFIX` or `$NPM_PREFIX` and then falling back to `$PREFIX`. But: There also need to be a way to disable the fallback legacy `$PREFIX` so that it can be ignored, if it's set by something else, so it [won't disrupt the logic](https://github.com/npm/libnpmconfig/blob/e9b201772fc461dd66b0b65a768e1f888273c302/index.js#L95) of `npm`. This way `nvm` could <s>set</s> _work_ with `$NODE_PREFIX` or `$NPM_PREFIX` instead of `$PREFIX` and Termux could safely continue to use `$PREFIX`, things wouldn't break.

Ideally, `npm` gets patched and released with this change for all versions that are shipped with `node` in the LTS channel as far back as possible. It would be nice to see this happen for npm 9, 8 and 6.

Related issues: 

- https://github.com/nvm-sh/nvm/issues/1645
- https://github.com/termux/termux-app/issues/2096

cc @ljharb @agnostic-apollo @goodevilgenius @GameKyuubi

**Comment by ljharb** [comment]
To be clear, nvm doesn't _set_ PREFIX, it merely checks it - because if it's set, then _npm_ will look outside of nvm's node directory for global modules.

**Reply by martin-braun** [reply]
Thanks for clarifying it. Guess I misinterpreted it, although the fix should still work.

**Comment by ljharb** [comment]
Since npm already looks at NPM_CONFIG_PREFIX, for a long time now, it would be a pretty small breaking change to just drop PREFIX entirely as an env var (even if it was kept in config as “prefix”).
